Added HTTP status code check feature > for the installation #18294

HRIT-Florian opened the pull request:
The security file check at the installation could be skipped, if the HTTP status code is not 401 (HTTP authentication required) which means the installation ist http access protected.

I think some web companys like me did work on http access protected development sites. The installation anyway take some steps, so, let's make the life easier and let's skip this if it is not needed.

For this, I have opened this pull request, where the http chek is now implemented.

### Summary of Changes
**Added HTTP status code check,**
for skipping security file if website is http 401 authentication required, which means http access protected.
https://github.com/joomla/joomla-cms/pull/18295/commits/70173fb9409d106c4b9790d87a345c4b2fa4f835

**Outsoureced functionality for the checks to two new methods**
called checkHostSecurity() and checkSecurityFile(). Did also renamed some names called remoteDBFile to securityFile, because i think it is more understandable. Because we are at the InstallationModelDatabase so think its clear which file and it's more associated to the method names. And some improvements in the check order logic. Hope the community like it ;)

The new methodes are called during initialise()

    // Save host checks
    if ($this->checkHostSecurity($options) === false)
    {
        if ($this->checkSecurityFile() === false)
        {
                return false;
        }
    }


### Testing Instructions
- Create a new installation site which is behind an directory protection like http access authentication.
- Go through the installation steps and you shouldn't see the security file check notification.
- And check again, but this time without the http directory protection and the security check should work, the notification should be shown as known ;)

### Documentation Changes Required
https://docs.joomla.org/J3.x:Secured_procedure_for_installing_Joomla_with_a_remote_database

Changes below in ***bold italic*** ↘ 

#### English New Documenation Part:
Starting with Joomla! 3.7.4 the Joomla! Security Strike Team (JSST) implemented additional security checks in the install application in order to protect your web hosting accounts from being overtaken by a remote attacker. In case your database is not on the same server as your website ***and your website is not through a HTTP authentication protected,*** we require an extra check that makes sure you are the owner of the website. 

...

#### German New Documenation Part:
Beginnend mit Joomla! 3.7.4 hat das Joomla! Security Strike Team (JSST) weitere Sicherheitsprüfungen in die Installations-Anwendung eingeführt, um Ihre Webhosting-Konten vor der Übernahme von einem entfernten Angreifer zu schützen. Im Falle, dass die genutzte Datenbank sich nicht auf dem gleichen Server befindet ***und die Website nicht durch eine HTTP-Authentifizierung geschützt ist,*** wird eine separate Prüfung verlangt, um zu versichern, dass Du der Inhaber der Website bist.

...

HRIT-Florian committed:
Added HTTP status code check,
for skipping security file if website is http 401 authentication required, which means http access protected.

brianteeman commented:
To me it looks like you haven't understood the purpose of the security check and the vulnerability it is protecting against.

HRIT-Florian commented:
Hi @brianteeman as i know, the security check is there to protect the installation from outside, right?

brianteeman commented:
No it's much more than that

HRIT-Florian commented:
Of course @brianteeman , and there are also other checks. But this PR did just effect the dbhost check. And it would be very exotic if we check also the dbhost if the site ist http protected. So, why shouldn't we simplify it on protected installations?

HRIT-Florian commented:
Did now also added the changes for the documentation in my PR comment. Documentation in English and German if this help. If wished, we could write it to the official documentation site: https://docs.joomla.org/J3.x:Secured_procedure_for_installing_Joomla_with_a_remote_database

brianteeman commented:
@SniperSister one for JSST to review

ghost commented:
corrected Typo in Title.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18295">issues.joomla.org/tracker/joomla-cms/18295</a>.</sub>

zero-24 commented:
@Didldu-Florian To me the idea looks ok but can we move that checks (all three) to a separate private method? That would keep the code simpler to understand as this checks should not all placed in the `initialise` method so we keep that small and simple to read. Thanks.

HRIT-Florian committed:
Merge branch 'staging' of github.com:joomla/joomla-cms into staging

HRIT-Florian committed:
Added checkSecurityFile and checkHostSecurity method

HRIT-Florian committed:
Moved remoteDbFileTest functionality to new separate mehtod checkSecurityFile()

HRIT-Florian committed:
Moved remote db_host and http_host to new separate mehtod checkHostSecurity()
Improved check logic !$shouldCheckLocalhost

HRIT-Florian committed:
Renamed remoteDbFile to securityFile

HRIT-Florian commented:
@zero-24 I'm glad you're okay with this. Did outsoureced the functionality for the checks to two new methods called checkHostSecurity() and checkSecurityFile(). Did also renamed some names called remoteDBFile to securityFile, because i think it is more understandable. Because we are at the InstallationModelDatabase so think its clear which file and it's more associated to the method names. And some improvements in the check order logic. Hope the community like it ;)

The new methodes are called during initialise()

    // Save host checks
    if ($this->checkHostSecurity($options) == false)
    {
            $this->checkSecurityFile();
    }

Hope it is more clear

HRIT-Florian committed:
Renamed remoteDbPath to $securityFilePath

zero-24 commented:
This does not work as expected. The check can be skipped. The method `initialise` needs to return false in case of an error / not matching file ;)

Didldu-Florian commented on installation/model/database.php:
```diff
@@ -80,6 +80,153 @@ public static function getUserId()
+		$uri = JUri::getInstance();
+
+		// HTTP Status Code
+		$statusCode = JHttpFactory::getHttp()->get((string) $uri)->code;
```
@zero-24 sorry, can't repoduce your concerns. Did tested again and it works. By the, way, the method initialise returns JDatabaseDriver or false on failure. Same as it worked before my PR.
Can you describe your concern in more detail? 

Please try hardset `$statusCode = 401;`  or `$statusCode = 200;` and be sure you are on localhost. 

HTTP 200 >The securityFileCheck will be required. 
HTTP 401 >The securityFileCheck will be skipped.

HRIT-Florian commented:
@zero-24 sorry, can't repoduce your concerns. Did tested again and it works. By the, way, the method `initialise` returns JDatabaseDriver or false on failure. Same as it worked before my PR.
Can you describe your concern in more detail? 

Please try hardset `$statusCode = 401;`  or `$statusCode = 200;` at [line 216](https://github.com/joomla/joomla-cms/pull/18295/files/01edd24b389060223ecab830449ffd4967c92da4#diff-7acfc2b93c1bae5b5114fc39c2c715ffR216) and be sure db_host is not localhost. 

HTTP 200 >The securityFileCheck will be required. 
HTTP 401 >The securityFileCheck will be skipped.

zero-24 commented:
@Didldu-Florian yes that part is working (401 aka .htaccess)

but when the check should be done you can just hit "install" and it starts doing things. (even that it stop working on another issue)

Basically that check in the `initialise` method should make sure that the validation fails and the user can't go any step forward. For that reason `initialise` needs to return false when the checks did not match on a host where we require to check it ;)

HRIT-Florian committed:
Added missing and required return for checkSecurityFile method

HRIT-Florian commented:
@zero-24 Oh dear! You are right, that was lost during outsourcing methods. My fault.
Did fixed it with https://github.com/joomla/joomla-cms/pull/18295/commits/c0a90ddc309b5bd62a9f1cf4d4554a7faef21b85 Now the logic is same as before on return.

zero-24 commented:
@Didldu-Florian it it just allowed to return false in case there is a error ;)

Else you can't move forward as the method return void just before we try to setup the database ;)

```php
// Save host checks
if ($this->checkHostSecurity($options) === false)
{
        if ($this->checkSecurityFile() === false)
        {
                return false;
        }
}
```

This should do the trick 😄

HRIT-Florian committed:
Improved return to return false just in case of error

HRIT-Florian commented:
@zero-24 Thanks, done ;)

Quy commented on installation/model/database.php:
```diff
@@ -80,6 +80,153 @@ public static function getUserId()
+	 *
+	 * @return  boolean  true if all checks have been passed
+	 *
+	 * @since   3.8.1
```
Please replace `@since 3.8.1` to `@since __DEPLOY_VERSION__` so it will be set automatically. Also below.

zero-24 committed:
just adding the missing =

zero-24 commented:
I have just added the missing= for the typesafe check. Looks good from here than. (the issue @quy mention should be fixed too)

HRIT-Florian committed:
Added since __DEPLOY_VERSION__ to both new methods

HRIT-Florian commented:
@zero-24, @Quy  `__DEPLOY_VERSION__` added ;)

Quy commented on installation/model/database.php:
```diff
@@ -80,6 +80,153 @@ public static function getUserId()
+	/**
+	 * Checks the remote db_host and http_host for security reason
+	 *
+	 * @param   object  &$options  The options referenz to use for configuration.
```
spelling `referenz`

HRIT-Florian replied:
@Quy done ;)

HRIT-Florian committed:
Fixed spelling in method doc block

zero-24 commented:
:+1: thanks @Didldu-Florian

bembelimen commented on installation/model/database.php:
```diff
@@ -79,6 +79,153 @@ public static function getUserId()
+		else
+		{
+			return true;
+		}
```
This should probably a return one liner:

```return in_array($options->db_host, $localhost) || $statusCode == 401```

brianteeman commented on installation/model/database.php:
```diff
@@ -79,6 +79,153 @@ public static function getUserId()
+		// HTTP Status Code
+		$statusCode = JHttpFactory::getHttp()->get((string) $uri)->code;
+
+		// Check the security file if the db_host is not save and HTTP status code is not 401 (HTTP authentication required)
```
please change "save" to either "safe" or "secure"

HRIT-Florian committed:
Update database.php

https://github.com/joomla/joomla-cms/pull/18295/files/52aa2c90842c571cc70ac2a8c653f8a7dc4c92c8#r204232543
https://github.com/joomla/joomla-cms/pull/18295/files/52aa2c90842c571cc70ac2a8c653f8a7dc4c92c8#r204232812

HRIT-Florian committed:
Merge branch 'staging' into staging

Quy commented:
Replaced by PR #26840<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18295">issues.joomla.org/tracker/joomla-cms/18295</a>.</sub>
